src/year2024/day06.rs

```diff
@@ -0,0 +1,157 @@
+
+    loop {
+        func(GridCell {
+            index: (row_i, column_i),
+            direction: direction,
+        });
+        let (next_row_i_option, next_col_i_option) = match direction {
+            Direction::Up if row_i > 0 => (Some(row_i - 1), Some(column_i)),
+            Direction::Down => (Some(row_i + 1), Some(column_i)),
+            Direction::Right => (Some(row_i), Some(column_i + 1)),
+            Direction::Left if column_i > 0 => (Some(row_i), Some(column_i - 1)),
+            _ => (None, None),
+        };
+
+        // If there is no next grid value, we are outside of the grid bounds.
+        if next_row_i_option == None || next_col_i_option == None {
+            break false;
+        }
+
+        let (next_row_i, next_col_i) = (next_row_i_option.unwrap(), next_col_i_option.unwrap());
+
+        let grid_next_option = grid.get(next_row_i, next_col_i);
+
+        if grid_next_option == None {
+            break false;
+        }
+
+        let grid_next = grid_next_option.unwrap();
+
+        // If there is an obstacle, rotate the direction and loop again
+        if grid_next == &'#' {
+            direction = match direction {
+                Direction::Up => Direction::Right,
+                Direction::Right => Direction::Down,
+                Direction::Down => Direction::Left,
+                Direction::Left => Direction::Up,
+            };
+            continue;
+        }
+
+        // Check if we've already navigated to this exact index.
+        // If we've already been to an index with the same direction, it means
+        // the guard is stuck in an infinite loop.
+        let is_new = path_steps.insert(GridCell {
+            index: (next_row_i, next_col_i),
+            direction: direction,
+        });
+
+        if is_new == false {
+            break true;
+        }
+
+        (row_i, column_i) = (next_row_i, next_col_i);
+    }
+}
+
+pub fn run() {
+    let grid = process_file("input/year2024/day06.txt");
+
+    let initial_cords = find_initial_cords(&grid);
+
+    let mut part1_grid = grid.clone();
+    let mut part2_grid = grid.clone();
+    let mut part1_count = 0;
+    let mut part2_count = 0;
+
+    process_grid(
+        &grid,
+        initial_cords,
+        &mut |GridCell {
+                  index: (row_i, col_i),
+                  ..
+              }| {
+            let grid_next_option = part1_grid.get(row_i, col_i);
+
+            if grid_next_option == Some(&'X') {
+                return;
+            }
+
+            part1_count += 1;
+            part1_grid[(row_i, col_i)] = 'X';
+
+            // Set one obstacle
+            part2_grid[(row_i, col_i)] = '#';
+            // Test
+            let result = process_grid(&part2_grid, initial_cords, &mut |GridCell { .. }| {});
+            if result == true {
+                part2_count += 1;
+            }
+            // Remove the placed obstacle
+            part2_grid[(row_i, col_i)] = '.';
```

Comment:
I'd be interested in re-visiting part 2 at some point. The complexity is O(n^2) which is unideal and definitely causes a noticeable slowdown. There are probably some optimizations that can be made to not run tests in certain cases or simplify the tests that do run.

Reply:
I think it's inevitable that the worst-case scenario for this will be O(n^2). Given each step in the guards path must individually be searched, that is just the limit.

The code could see a few optimizations in the future for speed:
- Parallelizing each temporary obstacle
- Tracking the previous space to remove redundant processing. Right now the obstacle checks start at the initial coordinates which results in extra processing.